docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,640 @@
+
+A Vite Module Runner allows running any code by processing it with Vite plugins first. It is different from `server.ssrLoadModule` because the runtime implementation is decoupled from the server. This allows library and framework authors to implement their own layer of communication between the Vite server and the runner. The browser communicates communicates with its corresponding environment using the server Web Socket and through HTTP requests. The SSR Module runner can directly do function calls to process modules. Other environments could run modules connecting to an edge runtime like workerd, or a Worker Thread in the same node process as Vitest does.
+
+All these environments share Vite's HTTP server, middlewares and Web Socket. The resolved config and plugins pipeline are also shared, but plugins can use `apply` so its hooks are only called for certain environments. The environment can also be accessed inside hooks for fine grained control.
+
+![Vite Environments](../images/vite-environments.svg)
+
+## Using environments in the Vite server
+
+A Vite dev server exposes two environments by default named `'browser'` and `'ssr'`. The browser environment runs in client apps that have imported the `/@vite/client` module. The SSR environment runs in the same runtime as the Vite server (for example, in node) and allows application servers to be used to render requests during dev with full HMR support. We'll discuss later how frameworks and users can create and register new environments.
+
+The available environments can be accessed using the `server.environments` read only array:
+
+```js
+server.environments.forEach((environment) => log(environment.name))
+```
+
+Each environment can also be accessed through its name:
+
+```js
+server.environment('browser').transformRequest(url)
+```
+
+An environment is an instance of the `ModuleExecutionEnvironment` class:
+
+```ts
+class ModuleExecutionEnvironment {
+  /**
+   * Unique identifier for the environment in a Vite server.
+   * By default Vite exposes 'browser' and 'ssr' environments.
+   * The ecosystem has concensus on other environments, like 'workerd'.
+   */
+  name: string
+  /**
+   * Communication channel to send and receive messages from the
+   * associated module runner in the target runtime.
+   */
+  hot: HMRChannel | null
+  /**
+   * Graph of module nodes, with the imported relationship between
+   * processed modules and the cached result of the processed code.
+   */
+  moduleGraph: ModuleGraph
+  /**
+   * Trigger the execution of a module using the associated module runner
+   * in the target runtime.
+   */
+  run: ModuleRunFunction
+  /**
+   * Resolved config options for this environment. Options at the server
+   * global scope are taken as defaults for all enviroments, and can
+   * be overriden (resolve conditions, external, optimizedDeps)
+   */
+  config: ResolvedEnvironmentConfig
+
+  constructor({ name, hot, run, config }: ModuleEnvironmentOptions)
```

Comment:
Should it receive a server? How does it access `pluginContainer` in `transformRequest`?

Reply:
Corrected 👍🏼